rust/protocol/src/curve.rs

```diff
@@ -280,41 +347,118 @@ impl TryFrom<&[u8]> for PrivateKey {
     }
 }
 
-#[derive(Copy, Clone)]
+impl subtle::ConstantTimeEq for PrivateKey {
+    fn ct_eq(&self, other: &PrivateKey) -> subtle::Choice {
+        if self.key_type() != other.key_type() {
+            return 0.ct_eq(&1);
+        }
+        self.key_data().ct_eq(other.key_data())
+    }
+}
+
+impl PartialEq for PrivateKey {
+    fn eq(&self, other: &PrivateKey) -> bool {
+        bool::from(self.ct_eq(other))
+    }
+}
+
+impl Ord for PrivateKey {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.key_type() != other.key_type() {
+            return self.key_type().cmp(&other.key_type());
+        }
+        constant_time_cmp(self.key_data(), other.key_data())
+    }
+}
+
+impl PartialOrd for PrivateKey {
+    fn partial_cmp(&self, other: &PrivateKey) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl fmt::Debug for PrivateKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            // Do not print out private keys in debug logs.
+            "PrivateKey {{ key_type={:?}, serialize=<...> }}",
+            self.key_type(),
+        )
+    }
+}
+
+/// A matching public and private key pair which can be used to encrypt and sign messages.
+///
+///```
+/// # fn main() -> libsignal_protocol::error::Result<()> {
+/// use libsignal_protocol::{KeyPair, PrivateKey, PublicKey};
+/// use rand::rngs::OsRng;
+/// use std::convert::TryInto;
+///
+/// let alice = KeyPair::generate(&mut OsRng);
+/// assert!(alice == alice.clone());
+/// let bob = KeyPair::generate(&mut OsRng);
+/// assert!(alice != bob);
+///
+/// // The first byte of `alice_zero` is 0, less than `bob_one` with 1, lexicographically:
+/// let mut alice_bytes: [u8; 32] = alice.public_key.public_key_bytes()?.try_into().unwrap();
+/// alice_bytes[0] = 0;
+/// let alice_zero = PublicKey::from_djb_public_key_bytes(alice_bytes.as_ref())?;
+///
+/// let mut bob_bytes: [u8; 32] = bob.public_key.public_key_bytes()?.try_into().unwrap();
+/// bob_bytes[0] = 1;
+/// let bob_one = PublicKey::from_djb_public_key_bytes(bob_bytes.as_ref())?;
+/// assert!(bob_one != bob.public_key);
+/// assert!(alice_zero < bob_one);
+/// assert!(alice_zero != bob_one);
+///
+/// // The same sort of thing can be done with private keys:
+/// let mut alice_private_bytes: [u8; 32] = alice.private_key.serialize().try_into().unwrap();
+/// alice_private_bytes[0] = !alice_private_bytes[0];
+/// let alice_flipped = PrivateKey::deserialize(alice_private_bytes.as_ref())?;
+/// assert!(alice_flipped != alice.private_key);
+///
+/// # Ok(())
+/// # }
+///```
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
```

Comment:
I don't think this needs to be Ord, actually. Eq is reasonable.

Reply:
Got it! Fixed!